### ISSUE TITLE
enable get the 'mathematical' gradient of output w.r.t neural network parameters

### ISSUE THREAD

**Issue by Erichliu00** [tensorflow/tensorflow#14954]
In some situation, it is necessary to get the 'mathematical' gradient of output w.r.t neural network parameters. 

For example, suppose I have a neural network `out = f(s)`, where `s` is a batch of input with shape`[None, dim_s]`, while out is a scaler, `f` is simply a MLP. With `tf.gradient(out, tf.trainable_variables())` I can get gradient of out w.r.t neural network parameters of f, which is a list of gradient. Now, I have two different batch of `s`: `s1` and `s2`, then we can get two different the above gradients `G1` and `G2`. It seems that it is impossible to compute cosine between `G1` and `G2` using current tensorflow? Do I need to flatten both gradients first? Do `G1` and `G2` are the usual gradient in math?

**Comment by aselle** [comment]
It's really difficult to follow what you are asking. TensorFlow computes gradients w.r.t. to scalar functions. I.e. usually you do tf.gradients on a loss which is a scalar (and if there is a batch, it will be some reduction across that batch too). If you give it a vector function it implicitly sums it first.   I'd suggest you setup an example with concrete things and make sure it is giving what you expect.

**Comment by Erichliu00** [comment]
Hi, @aselle Thanks. I have changed problem to scaler output to simplifier statement.

An example: 
```
import tensorflow as tf
import numpy as np

batch_1=np.random.normal(0,1, [2, 3])
batch_2=np.random.normal(0,1, [2, 3])
x = tf.placeholder(tf.float32, shape=(None, 3))

predictions = tf.layers.dense(x, 4, tf.tanh,
                              kernel_initializer=tf.random_normal_initializer(
                                  stddev=np.sqrt(1 / 100)))

opt = tf.train.GradientDescentOptimizer(learning_rate=0.01)
gradient_step = opt.compute_gradients(predictions, tf.trainable_variables())

sess=tf.Session()
sess.run(tf.global_variables_initializer())
gradients_1 = sess.run(gradient_step, feed_dict={x: batch_1})
gradients_2 = sess.run(gradient_step, feed_dict={x: batch_1})
```


Then both `gradients_1` and `gradients_2` will look something like:

> [(array([[-0.24431723, -0.25037131, -0.24526665, -0.25040874],
>        [-1.04525471, -1.03707826, -1.04750621, -1.04078865],
>        [ 1.6583091 ,  1.62776244,  1.6609472 ,  1.63547981]], dtype=float32), array([[-0.11156642, -0.04291356, -0.11599004,  0.12397685],
>        [-0.06834391, -0.00486498, -0.14670192, -0.01094443],
>        [ 0.04650469, -0.14726   , -0.10059511,  0.12041415]], dtype=float32)), (array([ 1.98409033,  1.96457601,  1.98815179,  1.97203505], dtype=float32), array([ 0.,  0.,  0.,  0.], dtype=float32))]

with shape as following:
> gradients_1[0][0].shape
> (3, 4)
> gradients_1[0][1].shape
> (3, 4)
> gradients_1[1][0].shape
> (4,)
> gradients_1[1][1].shape
> (4,)

Given the gradient is of such kind of shape, I think there is a gap between tensorflow gradient and the usual mathematical gradient(which usually is a vector or matrix), so it makes some applications, such as compute the cosine between `gradients_1` and `gradients_2` seems intractable.

**Comment by facaiy** [comment]
Hi, @Erichliu00 . 

You build the function: `prediction = y = tanh(w x + b)` and calculate gradients as

```python
gradient_step = opt.compute_gradients(predictions, tf.trainable_variables())
```

here `tf.trainable_variables()` is `[w, b]`, so we get: `dy / dw` and `dy / db`, the corresponding shapes are (3, 1) and (1,). I think everything works well.

Since you mentioned usual mathematical gradient, I guess that you might expect `dy / dx`, like:

```python
import tensorflow as tf
import numpy as np

batch_1=np.random.normal(0,1, [2, 3])
batch_2=np.random.normal(0,1, [2, 3])
x = tf.placeholder(tf.float32, shape=(None, 3))

predictions = tf.layers.dense(x, 1, tf.tanh,
                              kernel_initializer=tf.random_normal_initializer(
                                  stddev=np.sqrt(1 / 100)))

grad_by_x = tf.gradients(predictions, x)[0]

with tf.Session() as sess:
    sess.run(tf.global_variables_initializer())
    gradients_1 = sess.run(grad_by_x, feed_dict={x: batch_1})
    gradients_2 = sess.run(grad_by_x, feed_dict={x: batch_2})
    print("gradients_1: shape: {}, value:\n{}".format(gradients_1.shape, gradients_1))
    print("gradients_2: shape: {}, value:\n{}".format(gradients_2.shape, gradients_2))
```

```bash
~/Downloads ❯❯❯ python test.py
gradients_1: shape: (2, 3), value:
[[-0.00683434  0.2317003  -0.04384441]
 [-0.0068151   0.23104796 -0.04372097]]
gradients_2: shape: (2, 3), value:
[[-0.00593388  0.20117244 -0.03806766]
 [-0.00684012  0.23189643 -0.04388152]]
```

I wish the explanation could be useful.

**Comment by Erichliu00** [comment]
Hi @facaiy , thanks.  Indeed I want to get the gradient of output `predictions ` w.r.t neural network parameters `tf.trainable_variables `, however, my concern is that `tf.trainable_variables ` will contains some variables with shape like `3x2` then the gradient w.r.t such variable denoted as `g1` and `g2` will be of the same shape `3x2`, then my concerns are: 1. are the gradient `g1` and `g2` averaged over batch input or summed over batch, mathematically, seems we should get a averaged gradient. 2  how can we compute cosine between `g1` and `g2` given their shapes are not a vector.

```python
import tensorflow as tf
import numpy as np

batch_1=np.random.normal(0,1, [2, 3])
batch_2=np.random.normal(0,1, [2, 3])
x = tf.placeholder(tf.float32, shape=(None, 3))

out = tf.layers.dense(x, 2, tf.tanh,
                              kernel_initializer=tf.random_normal_initializer(
                                  stddev=np.sqrt(1 / 100)))
predictions = tf.layers.dense(out, 1, tf.tanh,
                              kernel_initializer=tf.random_normal_initializer(
                                  stddev=np.sqrt(1 / 100)))

opt = tf.train.GradientDescentOptimizer(learning_rate=0.01)
gradient_step = opt.compute_gradients(predictions, tf.trainable_variables())

sess=tf.Session()
sess.run(tf.global_variables_initializer())
gradients_1 = sess.run(gradient_step, feed_dict={x: batch_1})
gradients_2 = sess.run(gradient_step, feed_dict={x: batch_2})
```

**Comment by aselle** [comment]
gradients_1 and gradients_2 are the same shape. So just multiply them elementwise and sum the elements. You want to see if the gradients agree in direction i.e. the cosine, right? That should work, but perhaps you could give more background on what you are trying to accomplish. I think you might want to ask on Stack Overflow, as this seems to be more of a usage question rather than a bug in tensorflow.

**Comment by Erichliu00** [comment]
Yes, I want to see if the gradients agree in direction.

**Comment by aselle** [comment]
Sure, so just do
```
sum([np.dot(i,j) for i,j in zip(gradients_1, gradients_2)])
```
and that will be the cosine. The point is tf.gradients gives you an array entry for the gradient with respect to each discrete variable instead of a flattened one, because the optimizer needs to apply the gradients to each object in turn. .

**Comment by Erichliu00** [comment]
Hi @aselle , thanks, you are right, I should consider gradient of each variable to compute cosine and also should not expect a flatten gradient since variable has shape that is not flattened.  However, I still believe that if gradient is not the usual flattened gradient, then some computation may not applicable.
For example, suppose we have a neural network parameter `x` with shape `[2,4]`, then the gradient of output scaler w.r.t it is of the same shape `[2,4]`, then given four different batches of input, we can have four values of `x`, say `x0,x1,x2,x3= np.random.normal(0,1, [2,4])`,  how can we compute the variance of `x` given four observations?
